Apply guard knowledge to all simplifications

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114471
* #123436
* __->__ #123342

This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

lezcano committed:
Apply guard knowledge to all simplifications

This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

ezyang commented:
I thought of one reason this fails.

We run simplification when generating guards. Ordinarily, this is safe, because the only simplifications we used to do were replacements, but we separately generate guards for replacements.

However, now that you're applying guard knowledge to simplifications, we now have a funny problem where we are trying to simplify a guard "s0 < s1" in the process of emitting it, and you're like "Oh, this is True, because we already have a guard s0 < s1", which is NOT PERMISSIBLE.

We actually already have this problem for deferred runtime asserts (this is how I noticed it), but it's less noticeable because deferred runtime asserts are out of sight out of mind as long as you aren't violating them.

lezcano commented:
I figured that that could happen but
1. If this is the case, then I don't understand why https://github.com/pytorch/pytorch/pull/124217/ passes tests at all
2. We should be able to fix this by calling the simplification before adding the guard to the guards.

ezyang commented:
1. Ivan's PR is meant to be broadly semantics preserving? All he's doing is taking something that was repeatedly computed over and over again in a loop, and computing it once outside. Importantly, the guards/deferred_runtime_asserts fields don't change across loop iterations
2. Sure

lezcano commented:
Ah, right, he still just iterates over `for s in symbols_unbacked:`. I was mislead by the removal of `if s in self.var_to_val:`.

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3851,10 +3853,52 @@ def bound_sympy(self, expr: sympy.Expr, size_oblivious: bool = False) -> ValueRa
         return bound_sympy(expr, var_to_range)
 
+    @_lru_cache
+    def _get_axioms(self, symbols: Optional[Tuple["sympy.Symbol"]] = None) -> Tuple["sympy.Expr"]:
```
doc symbols would be helpful

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3851,10 +3853,52 @@ def bound_sympy(self, expr: sympy.Expr, size_oblivious: bool = False) -> ValueRa
     def _maybe_evaluate_static(
         self, expr: "sympy.Expr", *, unbacked_only: bool = False, compute_hint: bool = False,
-        expect_rational=True, size_oblivious: bool = False
+        expect_rational=True, size_oblivious: bool = False, axioms: Optional[Tuple[sympy.Expr]] = None
```
doc about appropriate use of axioms, basically what we discussed in the PR

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3624,10 +3624,12 @@ def issue_guard(guard: ShapeGuard) -> None:
+            if self._maybe_evaluate_static(guard.expr, axioms=previous_guards) is not None:
                 continue
             issue_guard(guard)
+            previous_guards += (guard.expr,)
```
Why not a list and append to it?

lezcano replied:
because we are caching on the args of the function, so we need it to be hashable

ezyang replied:
This seems suspicious. As axioms can be very large, you are essentially forcing an extremely large key in the cache entry, and you're also guaranteed to miss because the axioms list is always change on each iteration.

It seems better to just pass it empty axioms for all input calls, and rely on on-the-fly simplification when we were accumulating...

lezcano replied:
Yeah, I guess that this pass is not necessary now, right?

lezcano replied:
Well it is for now, as otherwise we would be emitting guards of the form `size()[0] <= 10`  and then `2 < = size()[0] <= 10`. We could remove the latter one and just keep the former + the lower bound, but I figured it'd be better to just keep the more compact form `2 <= size()[0] <= 10` for now, so leaving as-is + left a comment.

ezyang replied:
I think you gotta do a benchmark run then, this is potentially quadratic and that makes me very nervous

ezyang replied:
Also, value range refinements already nuke their corresponding guards, so you should never be seeing both of these anyways.

lezcano replied:
no, I removed passing the info iteratively in favour of not passing any axioms. This will not be as strong as it could be, but that's fine.

Regarding "nuking their guards", I guess that that was happening in this code. The life of a check is:
1. Try to prove that `size < 10` and fail
2. Refine the value range to be `sizes \in [2, 9]
3. In this code, prove that `size < 10` is not necessary via value ranges
4. Emit the guard `2 <= size <= 9`

If we remove 3, we end up with the extra guard. My point is that we could potentially delete 3 and just emit `2 <= size` if the lowerbound of `size` is 2. If it's anything else, don't emit it, as that means that it follows from the guards.

ezyang replied:
Ah. So, essentially, what I am hoping is that during the online process, if we see a guard that is 100% subsumed by the value range refinement, I hope we do not put the guard into the list at all. I don't remember if this is what we actually do or not, but that's my ideal situation.

lezcano replied:
That will happen, but it will not happen if the guard itself is the one that triggers the value range refinement. We could make it so that this is the case tho, in which case we could remove the simplifications from this for-loop an simply happily emit all the guards.

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]

lezcano committed:
Update on "Apply guard knowledge to all simplifications"


This was an oversight in a previous PR. We were just applying this
knowledge when the expression had an unbacked int

[ghstack-poisoned]
